### DISCUSSION TITLE
Parse request directly to Pydantic model using model_validate_json

### DISCUSSION THREAD

**Discussion by pepastach** [fastapi/fastapi#11279] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

class Payload(BaseModel):
    field: str

app = FastAPI()

@app.post("/check/")
async def check(payload: Payload):
    return payload
```

### Description

We profiled FastAPI requests using pyinstruments with thousands of requests to get some reasonable numbers and found that we spend almost 17% of time on `json.loads()` and another 17% on `ModelField.validate`. This makes me wonder why does FastAPI call Starlette's `json()` instead of just `body()`. The raw body string could then be passed to Pydantic's `model_validate_json()` right?

I'd appreciate if anyone who optimized their FastAPI app for performance could share their knowledge. We process relatively large requests (several kilobytes) so any performance improvement of this phase would be really appreciated.

<img width="767" alt="image" src="https://github.com/tiangolo/fastapi/assets/10232308/eb20461b-3327-4955-ab36-f754f338dd88">

Thank you!

### Operating System

macOS

### FastAPI Version

0.109.2

### Pydantic Version

2.5.3

### Python Version

3.11.5

**Comment by denis-samatov** [comment]
Yes — `model_validate_json()` can be faster for the narrow case “one JSON request body → one Pydantic model”, but FastAPI cannot use that path unconditionally.

FastAPI's body pipeline also has to support:

- multiple body parameters (for example, `a: ModelA, b: ModelB`), where the decoded object must be split before validation;
- embedded bodies, aliases, dependencies, forms/files, and non-JSON media types;
- FastAPI-style validation locations and errors;
- behavior that is consistent across Pydantic's JSON and Python validation modes.

That is why the general path parses the request first and then validates each body field. The same limitation was later confirmed in #9868.

For a measured hot path that accepts exactly one JSON model, you can opt into the direct Pydantic path explicitly:

```python
from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from pydantic import BaseModel, ValidationError

app = FastAPI()

class Payload(BaseModel):
    field: str

@app.post("/check/", response_model=Payload)
async def check(request: Request) -> Payload:
    raw = await request.body()
    try:
        return Payload.model_validate_json(raw)
    except ValidationError as exc:
        errors = [
            {**error, "loc": ("body", *error["loc"])}
            for error in exc.errors(include_url=False)
        ]
        raise RequestValidationError(errors, body=raw) from exc
```

The trade-off is that FastAPI no longer sees a declared request-body model, so you must add the request-body OpenAPI schema yourself (or implement the optimization in a custom `APIRoute`). I would only use this after profiling the complete endpoint: for many real APIs, dependency work, I/O, or response serialization dominates the JSON parsing cost.

*If this solves your issue, please consider marking this comment as the answer!*
